File: languages/php/tests/ConnectTimeoutDefaultTest.php
<?php

declare(strict_types=1);

namespace Revaly\Sdk\Tests;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\RequestOptions;
use PHPUnit\Framework\TestCase;
use Revaly\Sdk\RapClient;
use Revaly\Sdk\Testing\RapMockTransport;

/**
 * ADR-SDK-029 connect-timeout-default semantics: an omitted connectTimeout resolves to
 * the 10-second edge-ratified default, an explicit null opts out entirely, explicit
 * values pass through, and zero/negative values are rejected at construction.
 */
final class ConnectTimeoutDefaultTest extends TestCase
{
    /** Reads the CONNECT_TIMEOUT the client actually configured on its Guzzle wire. */
    private static function wiredConnectTimeout(RapClient $client): mixed
    {
        $prop = new \ReflectionProperty(RapClient::class, 'guzzle');
        /** @var GuzzleClient $guzzle */
        $guzzle = $prop->getValue($client);

        return $guzzle->getConfig(RequestOptions::CONNECT_TIMEOUT);
    }

    public function testOmittedResolvesToTheRatifiedDefault(): void
    {
        self::assertSame(10.0, RapClient::DEFAULT_CONNECT_TIMEOUT_SECONDS);

        $client = new RapClient(apiKey: 'sk-synthetic-test', transport: new RapMockTransport());

        self::assertSame(10.0, self::wiredConnectTimeout($client));
    }

    public function testExplicitNullOptsOutOfAnySdkConnectBound(): void
    {
        $client = new RapClient(
            apiKey: 'sk-synthetic-test',
            connectTimeout: null,
            transport: new RapMockTransport(),
        );

        // 0 is Guzzle's "no connect timeout" — the pre-ADR-029 unset behaviour.
        self::assertSame(0, self::wiredConnectTimeout($client));
    }

    public function testExplicitValuesPassThroughUnchanged(): void
    {
        $client = new RapClient(
            apiKey: 'sk-synthetic-test',
            connectTimeout: 3.0,
            transport: new RapMockTransport(),
        );

        self::assertSame(3.0, self::wiredConnectTimeout($client));
    }

    public function testZeroConnectTimeoutIsRejected(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        new RapClient(apiKey: 'sk-synthetic-test', connectTimeout: 0.0);
    }

    public function testNegativeConnectTimeoutIsRejected(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        new RapClient(apiKey: 'sk-synthetic-test', connectTimeout: -5.0);
    }
}

File: languages/go/tests/connect_timeout_default_test.go
// ADR-SDK-029 connect-timeout-default semantics: a zero-value Config resolves
// to the 10 s edge-ratified default, NoConnectTimeout opts out entirely, and
// negative values are still rejected. The dial-phase mapping itself (dialer
// bound → dial-phase *net.OpError → TransientFailure) is proven against a real
// wire in realwire_test.go.
package tests

import (
	"testing"
	"time"

	revaly "github.com/revaly-co/rap-sdk/languages/go"
)

func TestConnectDefaultConstantIsTheRatifiedValue(t *testing.T) {
	if revaly.DefaultConnectTimeout != 10*time.Second {
		t.Fatalf("DefaultConnectTimeout = %v, want 10s (ADR-SDK-029)", revaly.DefaultConnectTimeout)
	}
}

func TestNoConnectTimeoutIsAcceptedAsOptOut(t *testing.T) {
	cfg := revaly.Config{APIKey: syntheticKey, ConnectTimeout: revaly.NoConnectTimeout}
	if _, err := revaly.NewClient(cfg); err != nil {
		t.Fatalf("NewClient rejected NoConnectTimeout: %v", err)
	}
}

func TestNegativeConnectTimeoutIsStillRejected(t *testing.T) {
	cfg := revaly.Config{APIKey: syntheticKey, ConnectTimeout: -2 * time.Second}
	if _, err := revaly.NewClient(cfg); err == nil {
		t.Fatal("NewClient accepted a negative ConnectTimeout")
	}
}
